File: tradex/agent/model_registry.py
"""Model/provider registry for agent runtimes."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Awaitable, Callable

from ..config import AgentConfig
from ..config.agent_models import (
    ANTHROPIC_PROVIDER,
    CODEX_PROVIDER,
    AgentModelProfile,
    resolve_agent_model,
)
from .provider import LLMProvider, LLMProviderUnavailable

ProviderFactory = Callable[[AgentConfig, AgentModelProfile], LLMProvider]
ModelLister = Callable[[AgentConfig, AgentModelProfile], Awaitable[list[dict]]]


@dataclass(frozen=True)
class AgentModelProvider:
    """One registered model provider backend."""

    name: str
    factory: ProviderFactory
    list_models: ModelLister


class AgentModelRegistry:
    """Registry that resolves model config and instantiates providers."""

    def __init__(self) -> None:
        """初始化模型提供者注册表。"""
        self._providers: dict[str, AgentModelProvider] = {}

    def register(self, provider: AgentModelProvider) -> None:
        """注册一个模型提供者后端。"""
        self._providers[provider.name] = provider

    def resolve(self, config: AgentConfig) -> AgentModelProfile:
        """根据配置解析出对应的模型配置档案。"""
        return resolve_agent_model(config)

    def create_provider(self, config: AgentConfig) -> LLMProvider:
        """根据配置创建并返回对应的 LLM 提供者实例。"""
        profile = self.resolve(config)
        provider = self._providers.get(profile.provider)
        if provider is None:
            raise LLMProviderUnavailable(f"Unsupported agent provider: {profile.provider}")
        return provider.factory(config, profile)

    async def list_available_models(
        self, config: AgentConfig, *, provider_override: str | None = None,
    ) -> list[dict]:
        """列出指定提供者支持的所有可用模型。"""
        if provider_override:
            from ..config.agent_models import normalize_api_mode, normalize_model, normalize_reasoning_effort
            profile = AgentModelProfile(
                provider=provider_override,
                api_mode=normalize_api_mode(provider_override),
                model=normalize_model(provider_override, None),
                reasoning_effort=normalize_reasoning_effort(None),
                supports_reasoning=provider_override != "anthropic",
                requires_account_id=provider_override == "codex",
            )
        else:
            profile = self.resolve(config)
        provider = self._providers.get(profile.provider)
        if provider is None:
            raise LLMProviderUnavailable(f"Unsupported agent provider: {profile.provider}")
        return await provider.list_models(config, profile)


def default_agent_model_registry() -> AgentModelRegistry:
    """Build the built-in provider registry."""
    registry = AgentModelRegistry()

    def build_codex(config: AgentConfig, profile: AgentModelProfile) -> LLMProvider:
        """构建 Codex 提供者实例。"""
        from .providers.codex import CodexProvider

        return CodexProvider(config, profile)

    async def list_codex(config: AgentConfig, profile: AgentModelProfile) -> list[dict]:
        """列出 Codex 提供者支持的模型。"""
        from .providers.codex import CodexProvider

        return await CodexProvider(config, profile).list_models()

    def build_anthropic(config: AgentConfig, profile: AgentModelProfile) -> LLMProvider:
        """构建 Anthropic 提供者实例。"""
        from .providers.anthropic import AnthropicProvider

        return AnthropicProvider(config, profile)

    async def list_anthropic(config: AgentConfig, profile: AgentModelProfile) -> list[dict]:
        """列出 Anthropic 提供者支持的模型。"""
        from .providers.anthropic import AnthropicProvider

        return await AnthropicProvider(config, profile).list_models()

    registry.register(AgentModelProvider(CODEX_PROVIDER, build_codex, list_codex))
    registry.register(AgentModelProvider(ANTHROPIC_PROVIDER, build_anthropic, list_anthropic))
    return registry


DEFAULT_AGENT_MODEL_REGISTRY = default_agent_model_registry()

File: tradex/agent/model_registry.ts
/**
 * model_registry.ts — Convenience wrapper around AgentModel resolution + the
 * api_registry.
 *
 * Used by:
 *  - api/routes/agent.ts for listAvailableModels()
 *  - memory pipeline (via LLMProviderFactory)
 *
 * Internally delegates to resolveAgentModelFromConfig + getApiStream /
 * getApiListModels. memory's LLMChatClient.chat() shape is implemented here so
 * memory consumers don't have to know about the provider stream contract.
 */

import { AgentConfig } from "../config/index.js";
import {
  ANTHROPIC_PROVIDER,
  CODEX_PROVIDER,
  normalizeApiMode,
  normalizeModel,
  normalizeProvider,
  normalizeReasoningEffort,
} from "../config/agent_models.js";
import type { LLMChatClient, ChatResponse } from "./llm_client.js";
import type { AgentModel } from "./models.js";
import { resolveAgentModelFromConfig } from "./models.js";
import { getApiStream, getApiListModels } from "./api_registry.js";
import type { AgentContext, AgentModelDescriptor, TextContent } from "./core/types.js";
import { transformMessages } from "./core/transform-messages.js";
import { agentModelToDescriptor } from "./core/model-descriptor.js";

// Ensure built-in providers are registered
import "./providers/register.js";

export class LLMProviderUnavailable extends Error {}

/**
 * AgentModelRegistry — resolves config into AgentModel and provides
 * a simple chat client for the memory pipeline.
 */
export class AgentModelRegistry {
  /** Resolve config into an AgentModel value object. */
  resolve(config: AgentConfig): AgentModel {
    return resolveAgentModelFromConfig(config);
  }

  /**
   * Create an LLMChatClient from config. Wraps the provider stream into the
   * lightweight chat interface used by non-agentic consumers like the memory
   * pipeline.
   */
  createProvider(config: AgentConfig): LLMChatClient {
    const model = this.resolve(config);
    const descriptor = agentModelToDescriptor(model);
    const streamFn = getApiStream(descriptor.api);
    const apiKey = model.apiKey;
    return {
      name: descriptor.provider,
      model: descriptor.id,
      async chat({ system, messages, onDelta, signal }): Promise<ChatResponse> {
        const context: AgentContext = {
          systemPrompt: system ?? "",
          messages: transformMessages(messages, descriptor),
          tools: [],
        };
        const stream = streamFn(descriptor, context, {
          apiKey,
          signal,
        });
        // Forward text deltas to legacy onDelta callback while awaiting final result
        if (onDelta) {
          for await (const evt of stream) {
            if (evt.type === "text_delta" && evt.delta) {
              onDelta(evt.delta);
            }
          }
        }
        const message = await stream.result();
        const content = message.content
          .filter((c): c is TextContent => c.type === "text")
          .map((c) => c.text)
          .join("");
        return { content, message };
      },
    };
  }

  /** List available models for a provider. */
  async listAvailableModels(
    config: AgentConfig,
    providerOverride?: string | null,
  ): Promise<Array<Record<string, unknown>>> {
    let model: AgentModel;
    if (providerOverride) {
      const provider = normalizeProvider(providerOverride);
      const apiMode = normalizeApiMode(provider);
      const stub: AgentConfig = {
        ...config,
        provider,
        apiMode,
        model: normalizeModel(provider, null),
        reasoningEffort: normalizeReasoningEffort(null),
      };
      model = resolveAgentModelFromConfig(stub);
    } else {
      model = this.resolve(config);
    }
    const descriptor = agentModelToDescriptor(model);
    const listFn = getApiListModels(descriptor.api);
    return listFn(descriptor, { apiKey: model.apiKey });
  }
}

export const DEFAULT_AGENT_MODEL_REGISTRY = new AgentModelRegistry();

// Re-export the supported provider constants so callers don't have to import
// from config to know what's registered.
export { ANTHROPIC_PROVIDER, CODEX_PROVIDER };
